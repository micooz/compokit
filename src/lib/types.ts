import type { Note } from "./note";
import type { NoteArray } from "./note-array";

export enum ModeEnum {
  // major and minor
  NaturalMajor,
  HarmonicMajor,
  MelodicMajor,
  NaturalMinor,
  HarmonicMinor,
  MelodicMinor,
  // church mode
  Ionian,
  Dorian,
  Phrygian,
  Lydian,
  Mixolydian,
  Aeolian,
  Locrian,
  // jazz
  MajorBlues,
  MinorBlues,
  JazzMelodicMinor,
  JazzHarmonicMinor,
  Bebop,
  Diminished,
}

export type ModeNameOptions = { shortName?: boolean } & NoteToNameOptions;

export type ResolveToOptions = {
  algorithm: "closely-related-modes";
};

export enum ChordTypeEnum {
  Triad,
  Seventh,
  // Ninth,
  // Eleventh,
  // Thirteen,
}

export type ChordIsOptions = {
  checkInversion?: boolean;
  nodeIsOptions?: NoteIsOptions;
};

export type ChordJSON = {
  notes: string[];
  mode?: { key: string; type?: ModeEnum };
  step?: number;
};

export enum NoteEnum {
  C = 1,
  D,
  E,
  F,
  G,
  A,
  B,
}

export enum AccidentalEnum {
  Natural,
  NaturalWithSymbol,
  Sharp,
  Flat,
  DoubleSharp,
  DoubleFlat,
}

export type NoteToNameOptions = {
  transformAccidental?: boolean;
};

export type NoteIsOptions = {
  checkGroup?: boolean;
  checkAccidental?: boolean;
};

export type NoteType = string | Note;

export type Notes = string[] | Note[] | NoteArray;
