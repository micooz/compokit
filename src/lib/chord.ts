import { Note, NoteToNameOptions, Notes } from "./note";
import { NoteArray } from "./note-array";

export enum ChordTypeEnum {
  Triad,
  Seventh,
  // Ninth,
  // Eleventh,
  // Thirteen,
}

const chordQualityMap: Record<string, string> = {
  "m-m": "dim",
  "m-M": "m",
  "M-m": "", // the triad doesn't need an "M"
  "M-M": "aug",

  "m-m-m": "dim7",
  "m-m-M": "m7(♭5)",
  "m-M-m": "m7",
  "m-M-M": "mM7",
  "M-m-m": "7",
  "M-m-M": "maj7",
  "M-M-m": "aug7",
  "M-M-M": "augM7",
};

const degreeSeqToInversionAndRootIndex: Record<string, [number, number]> = {
  "3-3": [0, 0], // C E G
  "3-4": [1, 2], // E G C
  "4-3": [2, 1], // G C E
  "3-3-3": [0, 0], // C E G B
  "3-3-2": [1, 3], // E G B C
  "3-2-3": [2, 2], // G B C E
  "2-3-3": [3, 1], // B C E G
};

export class Chord {
  private _noteArr: NoteArray;

  private _root: Note;

  private _inversion = 0;

  constructor(notes: Notes) {
    this._noteArr = new NoteArray(notes);

    // TODO: supports structures containing omitted and added notes
    if (notes.length < 3) {
      throw new Error("chord must contain at least three note");
    }

    // dedup
    const dedup = this._noteArr.dedup();

    // intervals in pairs
    const intervals = dedup.intervals();

    // extract degrees
    const degrees = intervals.map((inter) => inter.degree.valueOf());

    // determine inversion and root index
    const [inversion, rootIndex] =
      degreeSeqToInversionAndRootIndex[degrees.join("-")] || [];

    if (inversion === undefined || rootIndex === undefined) {
      throw new Error(
        `cannot form a chord with these notes: ${this._noteArr.join()}`
      );
    }

    // root note
    const root = dedup.get(rootIndex);

    this._root = root;
    this._inversion = inversion;
  }

  private intervals() {
    return this._noteArr.intervals();
  }

  // TODO: fromAbbr
  fromAbbr(abbr: string) {
    //
  }

  inversion(ordinal = 1) {
    const backward = ordinal < 0;
    const notes = this._noteArr.clone().valueOf();

    for (let n = 1; n <= Math.abs(ordinal); n += 1) {
      if (backward) {
        const last = notes.pop()!;

        if (last.group) {
          last.group -= 1;
        }

        notes.unshift(last);
      } else {
        const first = notes.shift()!;

        if (first.group) {
          first.group += 1;
        }

        notes.push(first);
      }
    }

    return new Chord(notes);
  }

  inversionOrdinal() {
    return this._inversion;
  }

  rootPosition() {
    if (this._inversion === 0) {
      return this.clone();
    }

    return this.inversion(-this._inversion);
  }

  toAbbr(opts?: NoteToNameOptions) {
    const first = this._noteArr.get(0).name(opts);

    // C
    const root = this._root.name(opts);

    // chord in root position
    const inRootPosition = this.inversion(-this._inversion);

    // quality sequence
    const qualitySeq = inRootPosition
      .intervals()
      .map((quality) => {
        if (quality.isMinor()) {
          return "m";
        }
        if (quality.isMajor()) {
          return "M";
        }
        if (quality.isAugmented()) {
          return "aug";
        }
        if (quality.isDiminished()) {
          return "dim";
        }
      })
      .filter(Boolean)
      .join("-");

    // m/M/aug/dim(7/9/11/13)
    const quality = chordQualityMap[qualitySeq];

    if (quality === undefined) {
      throw new Error(
        `unable to determine the appropriate chord representation for: ${this.join()}`
      );
    }

    // (♭5)(♭9)(♭13)
    const accidentals = [].join("");

    // /G
    let by = "";

    if (root !== first) {
      by = `/${first}`;
    }

    // CmM7/G
    return `${root}${quality}${accidentals}${by}`;
  }

  notes() {
    return this._noteArr;
  }

  noteNames(opts?: NoteToNameOptions) {
    return this._noteArr.names(opts);
  }

  join(separator = " ", opts?: NoteToNameOptions) {
    return this._noteArr.join(separator, opts);
  }

  clone() {
    return new Chord(this._noteArr.valueOf());
  }
}
