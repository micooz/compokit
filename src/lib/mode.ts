import { Chord } from "./chord";
import { range, transformObject } from "./utils";
import { Interval } from "./interval";
import { Note } from "./note";
import { NoteArray } from "./note-array";
import {
  AccidentalEnum,
  ChordTypeEnum,
  Inter,
  ModeEnum,
  ModeNameOptions,
  NoteType,
  Notes,
} from "./types";

const majorMinorRelations: Record<number, ModeEnum> = {
  [ModeEnum.NaturalMajor]: ModeEnum.NaturalMinor,
  [ModeEnum.HarmonicMajor]: ModeEnum.HarmonicMinor,
  [ModeEnum.MelodicMajor]: ModeEnum.MelodicMinor,
  [ModeEnum.NaturalMinor]: ModeEnum.NaturalMajor,
  [ModeEnum.HarmonicMinor]: ModeEnum.HarmonicMajor,
  [ModeEnum.MelodicMinor]: ModeEnum.MelodicMajor,
};

const modeGroups: Record<string, Partial<Record<ModeEnum, string>>> = {
  "Major and Minor": {
    [ModeEnum.NaturalMajor]: "Natural Major",
    [ModeEnum.HarmonicMajor]: "Harmonic Major",
    [ModeEnum.MelodicMajor]: "Melodic Major",
    [ModeEnum.NaturalMinor]: "Natural Minor",
    [ModeEnum.HarmonicMinor]: "Harmonic Minor",
    [ModeEnum.MelodicMinor]: "Melodic Minor",
  },
  "Church Mode": {
    [ModeEnum.Ionian]: "Ionian", // same as ModeEnum.NaturalMajor
    [ModeEnum.Dorian]: "Dorian",
    [ModeEnum.Phrygian]: "Phrygian",
    [ModeEnum.Lydian]: "Lydian",
    [ModeEnum.Mixolydian]: "Mixolydian",
    [ModeEnum.Aeolian]: "Aeolian", // same as ModeEnum.NaturalMinor
    [ModeEnum.Locrian]: "Locrian",
  },
  Jazz: {
    // [ModeEnum.MajorBlues]: "Major Blues",
    // [ModeEnum.MinorBlues]: "Minor Blues",
    [ModeEnum.JazzMelodicMinor]: "Melodic Minor (Jazz)",
    [ModeEnum.JazzHarmonicMinor]: "Harmonic Minor (Jazz)",
    // [ModeEnum.Bebop]: "Bebop",
    // [ModeEnum.Diminished]: "Diminished",
  },
};

const chordTypeEnumToIndexes: Record<ChordTypeEnum, number[]> = {
  [ChordTypeEnum.Triad]: [1, 3, 5],
  [ChordTypeEnum.Seventh]: [1, 3, 5, 7],
  // [ChordTypeEnum.Ninth]: [1, 3, 5, 7, 9],
  // [ChordTypeEnum.Eleventh]: [1, 3, 5, 7, 9, 11],
  // [ChordTypeEnum.Thirteen]: [1, 3, 5, 7, 9, 11, 13],
};

const modeToIntervals = transformObject(
  {
    // major and minor
    [ModeEnum.NaturalMajor]: ["P1", "M2", "M3", "P4", "P5", "M6", "M7"],
    [ModeEnum.HarmonicMajor]: ["P1", "M2", "M3", "P4", "P5", "m6", "M7"],
    [ModeEnum.MelodicMajor]: ["P1", "M2", "M3", "P4", "P5", "m6", "m7"],
    [ModeEnum.NaturalMinor]: ["P1", "M2", "m3", "P4", "P5", "m6", "m7"],
    [ModeEnum.HarmonicMinor]: ["P1", "M2", "m3", "P4", "P5", "m6", "M7"],
    [ModeEnum.MelodicMinor]: ["P1", "M2", "m3", "P4", "P5", "M6", "M7"],
    // church mode
    [ModeEnum.Ionian]: ["P1", "M2", "M3", "P4", "P5", "M6", "M7"], // same as ModeEnum.NaturalMajor
    [ModeEnum.Dorian]: ["P1", "M2", "m3", "P4", "P5", "M6", "m7"],
    [ModeEnum.Phrygian]: ["P1", "m2", "m3", "P4", "P5", "m6", "m7"],
    [ModeEnum.Lydian]: ["P1", "M2", "M3", "A4", "P5", "M6", "M7"],
    [ModeEnum.Mixolydian]: ["P1", "M2", "M3", "P4", "P5", "M6", "m7"],
    [ModeEnum.Aeolian]: ["P1", "M2", "m3", "P4", "P5", "m6", "m7"], // same as ModeEnum.NaturalMinor
    [ModeEnum.Locrian]: ["P1", "m2", "m3", "P4", "d5", "m6", "m7"],
    // jazz
    [ModeEnum.MajorBlues]: ["P1", "M2", "m3", "M3", "P5", "M6"],
    [ModeEnum.MinorBlues]: ["P1", "m3", "P4", "d5", "P5", "m7"],
    [ModeEnum.JazzMelodicMinor]: ["P1", "M2", "m3", "P4", "P5", "M6", "M7"],
    [ModeEnum.JazzHarmonicMinor]: ["P1", "M2", "m3", "P4", "P5", "m6", "M7"],
    [ModeEnum.Bebop]: ["P1", "M2", "M3", "P4", "P5", "M6", "m7", "M7"],
    [ModeEnum.Diminished]: ["P1", "M2", "m3", "P4", "d5", "m6", "M6", "M7"],
  },
  (_, abbrs) => abbrs.map((abbr) => Interval.from(abbr))
);

export class Mode {
  private _noteArr: NoteArray;

  private _mode?: ModeEnum;

  constructor(notes: Notes, mode?: ModeEnum) {
    this._noteArr = Array.isArray(notes) ? new NoteArray(notes) : notes;

    if (this._noteArr.count() === 0) {
      throw new Error("mode must contains at least one note");
    }

    this._mode = mode;
  }

  static from(key: NoteType, mode: ModeEnum) {
    const intervals = modeToIntervals[mode];

    if (!intervals) {
      throw new Error(`unknown mode: ${mode}`);
    }

    const root = typeof key === "string" ? Note.from(key) : key;
    const notes = intervals.map((inter) => root.add(inter));

    // In harmonic minor, force set 7th note(which accidental is AccidentalEnum.Natural) to AccidentalEnum.NaturalWithSymbol.
    if (
      mode === ModeEnum.HarmonicMinor &&
      notes[6].accidental === AccidentalEnum.Natural
    ) {
      notes[6].accidental = AccidentalEnum.NaturalWithSymbol;
    }

    return new Mode(notes, mode);
  }

  static modes() {
    return Object.entries(modeGroups).map(([group, map]) => ({
      group,
      items: Object.entries(map).map(([key, value]) => ({
        label: value,
        value: Number(key) as ModeEnum,
      })),
    }));
  }

  static getKeys(mode: ModeEnum) {
    const allKeys = range("C", "B", "both").reduce<Note[]>(
      (acc, next) => acc.concat(next.valueOf()),
      []
    );

    // exclude double flat and double sharp key
    // exclude all flat and all sharp key
    const filteredKeys = allKeys.filter((key) => {
      const notes = Mode.from(key, mode).notes().valueOf();

      let count = 0;

      for (const note of notes) {
        if (note.isDoubleFlat() || note.isDoubleSharp()) {
          return false;
        }
        if (!note.isNature()) {
          count += 1;
        }
      }

      if (count === notes.length) {
        return false;
      }

      return true;
    });

    return NoteArray.from(filteredKeys);
  }

  key() {
    return this._noteArr.get(0)!;
  }

  type() {
    return this._mode;
  }

  chord(step: number, type: ChordTypeEnum) {
    const count = this._noteArr.count();

    if (step < 1 || step > count) {
      throw new Error(`step must be between 1 and ${count}, but got: ${step}`);
    }

    const indexes = chordTypeEnumToIndexes[type];

    if (!indexes) {
      throw new Error(`unknown chord type: ${type}`);
    }

    const notes = indexes.map((index) => {
      const idxAbs = step + index - 1;

      let idx = idxAbs % count;
      idx = idx < 1 ? idx + count : idx;

      const note = this._noteArr.get(idx - 1)!;

      if (note.group) {
        note.group += Math.floor((idxAbs - 1) / count);
      }

      return note;
    });

    return new Chord(notes, this.clone(), step);
  }

  chords(type: ChordTypeEnum) {
    const to = this.notes().count();

    const chords: Chord[] = [];

    for (let step = 1; step <= to; step += 1) {
      const chord = this.chord(step, type);
      chords.push(chord);
    }

    return chords;
  }

  transpose(direction: "low" | "high", inter: Inter) {
    if (direction === "low") {
      this._noteArr = this._noteArr.map((note) => note.minus(inter));
    }
    if (direction === "high") {
      this._noteArr = this._noteArr.map((note) => note.add(inter));
    }

    return this;
  }

  parallel() {
    if (!this.isMajor() && !this.isMinor()) {
      throw new Error("parallel() only works on major or minor mode");
    }

    const root = this._noteArr.root();
    const mode = majorMinorRelations[this._mode!];

    return Mode.from(root, mode);
  }

  relative() {
    if (!this.isMajor() && !this.isMinor()) {
      throw new Error("relative() only works on major or minor mode");
    }

    let root = this._noteArr.root();

    if (this.isMajor()) {
      root = root.minus("m3");
    }
    if (this.isMinor()) {
      root = root.add("m3");
    }

    const mode = majorMinorRelations[this._mode!];

    return Mode.from(root, mode);
  }

  dominant() {
    return Mode.from(this._noteArr.get(4)!, this._mode!);
  }

  subDominant() {
    return Mode.from(this._noteArr.get(3)!, this._mode!);
  }

  name(opts?: ModeNameOptions) {
    const root = this._noteArr.root().name(opts);

    let mode = "";

    if (this._mode === undefined) {
      mode = "Unknown Mode";
    } else {
      const map = Object.values(modeGroups).reduce((acc, next) => {
        return { ...acc, ...next };
      }, {});

      mode = map[this._mode]!;
    }

    if (opts?.shortName) {
      mode = mode
        .replace("Natural ", "")
        .replace("Harmonic ", "")
        .replace("Melodic ", "");
    }

    return `${root} ${mode}`;
  }

  notes() {
    return this._noteArr;
  }

  clone() {
    return new Mode(this._noteArr.clone(), this._mode);
  }

  private isMajor() {
    return (
      typeof this._mode !== "undefined" &&
      [
        ModeEnum.NaturalMajor,
        ModeEnum.HarmonicMajor,
        ModeEnum.MelodicMajor,
      ].includes(this._mode)
    );
  }

  private isMinor() {
    return (
      typeof this._mode !== "undefined" &&
      [
        ModeEnum.NaturalMinor,
        ModeEnum.HarmonicMinor,
        ModeEnum.MelodicMinor,
      ].includes(this._mode)
    );
  }
}
