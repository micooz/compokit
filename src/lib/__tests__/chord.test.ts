import { Chord } from "../chord";
import { Mode } from "../mode";
import { ModeEnum } from "../types";

test("Chord::new", () => {
  expect(() => new Chord([])).toThrow("chord must contain at least three note");
  expect(() => new Chord(["C", "D", "E"])).toThrow(
    "cannot form a chord with these notes: C D E"
  );

  expect(new Chord(["C", "E", "G"]).toAbbr()).toBe("C");
  expect(new Chord(["C", "Eb", "G"]).toAbbr()).toBe("Cm");
  expect(new Chord(["C", "Eb", "Gb"]).toAbbr()).toBe("Cdim");
  expect(new Chord(["C", "E", "G#"]).toAbbr()).toBe("Caug");
  expect(new Chord(["E", "G", "C"]).toAbbr()).toBe("C/E");
  expect(new Chord(["C", "E", "G", "B"]).toAbbr()).toBe("Cmaj7");
  expect(new Chord(["C", "E", "G", "Bb"]).toAbbr()).toBe("C7");
  expect(new Chord(["C", "Eb", "Gb", "Bb"]).toAbbr()).toBe("Cm7(â™­5)");

  expect(() => new Chord(["C", "E#", "G"]).toAbbr()).toThrow(
    "unable to determine the appropriate chord representation for: C E# G"
  );
});

test("Chord::fromAbbr", () => {
  //
});

test("Chord::fromJSON", () => {
  expect(Chord.fromJSON({ notes: ["C", "E", "G"] }).toAbbr()).toBe("C");
  expect(
    Chord.fromJSON({
      notes: ["C", "E", "G"],
      mode: { key: "C", type: ModeEnum.NaturalMajor },
    }).toAbbr()
  ).toBe("C");
});

test("Chord::toAbbr", () => {
  const chord = new Chord(["E", "G", "C"]);
  expect(chord.toAbbr()).toBe("C/E");
});

test("Chord::inverse", () => {
  const chord = new Chord(["F", "A", "C"]);
  expect(chord.inverse().notes().join()).toBe("A C F");

  // with group
  const chord2 = new Chord(["G4", "B4", "D5", "F5"]);
  expect(chord2.inverse(-1).notes().join()).toBe("F4 G4 B4 D5");
  expect(chord2.inverse(1).notes().join()).toBe("B4 D5 F5 G5");
  expect(chord2.inverse(2).notes().join()).toBe("D5 F5 G5 B5");
  expect(chord2.inverse(3).notes().join()).toBe("F5 G5 B5 D6");
});

test("Chord::inversion", () => {
  const chord = new Chord(["E", "G", "C"]);
  expect(chord.inversion()).toBe(1);
});

test("Chord::rootPosition", () => {
  const chord = new Chord(["E", "G", "C"]);
  expect(chord.rootPosition().notes().join()).toBe("C E G");

  const chord2 = new Chord(["C", "E", "G"]);
  expect(chord2.rootPosition().notes().join()).toBe("C E G");
});

test("Chord::notes", () => {
  const chord = new Chord(["E", "G", "C"]);
  expect(chord.notes().join()).toBe("E G C");
});

test("Chord::is", () => {
  const chord = new Chord(["E", "G", "C"]);

  expect(chord.is(new Chord(["C", "E", "G"]))).toBe(true);
  expect(chord.is(new Chord(["C", "E", "G"]), { checkInversion: true })).toBe(
    false
  );
});

test("Chord::resolveTo", () => {
  const format = (chord: Chord) => ({
    chord: chord.toAbbr(),
    mode: chord.mode()!.name(),
    step: chord.step(),
  });

  const chords = [
    new Chord(["G", "B", "D"], Mode.from("C", ModeEnum.NaturalMajor)),
    new Chord(["G", "B", "D", "F"], Mode.from("C", ModeEnum.NaturalMajor)),
    new Chord(["G", "Bb", "Db", "F"], Mode.from("F", ModeEnum.HarmonicMinor)),
  ];

  const snapshot = chords.map((chord) => {
    const to = chord
      .resolveTo({ algorithm: "closely-related-modes" })
      .map((chord) => format(chord));

    return {
      from: format(chord),
      to,
    };
  });

  expect(snapshot).toMatchSnapshot();

  // @ts-ignore
  expect(new Chord(["C", "E", "G"]).resolveTo({ algorithm: null })).toEqual([]);

  expect(() =>
    new Chord(["C", "E", "G"]).resolveTo({ algorithm: "closely-related-modes" })
  ).toThrow("this chord is not associated with a mode");
});

test("Chord::clone", () => {
  const chord = new Chord(["E", "G", "C"]);
  expect(chord.clone().is(chord)).toBe(true);
});

test("Chord::toJSON", () => {
  const chord = new Chord(
    ["E", "G", "C"],
    Mode.from("G", ModeEnum.NaturalMajor),
    4
  );
  expect(chord.toJSON()).toEqual({
    mode: {
      key: "G",
      type: 0,
    },
    notes: ["E", "G", "C"],
    step: 4,
  });
});
