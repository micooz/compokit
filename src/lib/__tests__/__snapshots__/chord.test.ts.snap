// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chord::resolveTo 1`] = `
[
  {
    "from": {
      "chord": "G",
      "mode": "C Natural Major",
      "step": undefined,
    },
    "to": [
      {
        "chord": "C",
        "mode": "C Natural Major",
        "step": 1,
      },
      {
        "chord": "C",
        "mode": "A Natural Minor",
        "step": 3,
      },
      {
        "chord": "C",
        "mode": "E Natural Minor",
        "step": 6,
      },
      {
        "chord": "C",
        "mode": "D Natural Minor",
        "step": 7,
      },
      {
        "chord": "C",
        "mode": "G Natural Major",
        "step": 4,
      },
      {
        "chord": "C",
        "mode": "F Natural Major",
        "step": 5,
      },
      {
        "chord": "Cm",
        "mode": "C Natural Minor",
        "step": 1,
      },
      {
        "chord": "Eb",
        "mode": "C Natural Minor",
        "step": 3,
      },
      {
        "chord": "Ab",
        "mode": "C Natural Minor",
        "step": 6,
      },
    ],
  },
  {
    "from": {
      "chord": "G7",
      "mode": "C Natural Major",
      "step": undefined,
    },
    "to": [
      {
        "chord": "C",
        "mode": "C Natural Major",
        "step": 1,
      },
      {
        "chord": "C",
        "mode": "A Natural Minor",
        "step": 3,
      },
      {
        "chord": "C",
        "mode": "E Natural Minor",
        "step": 6,
      },
      {
        "chord": "C",
        "mode": "D Natural Minor",
        "step": 7,
      },
      {
        "chord": "C",
        "mode": "G Natural Major",
        "step": 4,
      },
      {
        "chord": "C",
        "mode": "F Natural Major",
        "step": 5,
      },
      {
        "chord": "Cm",
        "mode": "C Natural Minor",
        "step": 1,
      },
      {
        "chord": "Eb",
        "mode": "C Natural Minor",
        "step": 3,
      },
      {
        "chord": "Ab",
        "mode": "C Natural Minor",
        "step": 6,
      },
    ],
  },
  {
    "from": {
      "chord": "Gm7(â™­5)",
      "mode": "F Harmonic Minor",
      "step": undefined,
    },
    "to": [
      {
        "chord": "Abaug",
        "mode": "F Harmonic Minor",
        "step": 3,
      },
      {
        "chord": "Ab",
        "mode": "Ab Harmonic Major",
        "step": 1,
      },
      {
        "chord": "Abm",
        "mode": "Eb Harmonic Major",
        "step": 4,
      },
      {
        "chord": "Cbaug",
        "mode": "Eb Harmonic Major",
        "step": 6,
      },
      {
        "chord": "Gbm",
        "mode": "Db Harmonic Major",
        "step": 4,
      },
      {
        "chord": "Ab",
        "mode": "Db Harmonic Major",
        "step": 5,
      },
      {
        "chord": "Ab",
        "mode": "C Harmonic Minor",
        "step": 6,
      },
      {
        "chord": "Gb",
        "mode": "Bb Harmonic Minor",
        "step": 6,
      },
    ],
  },
]
`;
